src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationHealthCheck.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using System;
+using System.Threading.Tasks;
+using System.Threading;
+
+namespace Microsoft.Extensions.Configuration.AzureAppConfiguration
+{
+    internal class AzureAppConfigurationHealthCheck : IConfigurationHealthCheck
+    {
+        private AzureAppConfigurationProvider _provider = null;
+
+        internal void SetProvider(AzureAppConfigurationProvider provider)
+        {
+            _provider = provider ?? throw new ArgumentNullException(nameof(provider));
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            if (_provider == null)
+            {
+                return HealthCheckResult.Unhealthy("Configuration provider is not set.");
+            }
+
+            if (!_provider.LastSuccessfulAttempt.HasValue)
+            {
+                return HealthCheckResult.Unhealthy("The initial load is not completed.");
+            }
+
+            if (_provider.LastFailedAttempt.HasValue &&
+                _provider.LastSuccessfulAttempt.Value < _provider.LastFailedAttempt.Value)
```

Comment:
If the refresh occurs frequently (eg. every a few seconds), we shouldn't report unhealthy just because last refresh failed for whatever reason. It causes unnecessary panic/alarm.

Reply:
I wouldn't say reporting unhealthy should be tied to the frequency of refresh. After all, follow up attempts on failure are not based on the configured frequency, they are based on our backup strategy. 

Typically, the orchestrator that is monitoring the health of the app is the one to take an action on whether the app is healthy or not and is the one to protect against unnecessary action.

If any change is made, I would expect an option, either a failure duration or failure count to represent how long the app configuration provider has failed to reach the service which leads to unhealthy. Otherwise, anything we pick would be arbitrary.